file/cache.lisp

```diff
@@ -36,55 +44,93 @@
+      (setf (file-%length stream) (1+ (file-%position stream))))
+    (setf (aref (file-%buffer stream) block-offset) byte
+          (file-%dirty-bit stream) t)
+    (incf (file-%position stream))))
 
 (defmethod mezzano.gray:stream-read-byte ((stream file-cache-stream))
-  (assert (member (direction stream) '(:input :io)))
-  (prog1 (if (>= (file-position* stream) (file-length* stream))
-             :eof
-             (aref (buffer stream) (file-position* stream)))
-    (incf (file-position* stream))))
-
-(defmethod mezzano.gray:stream-write-sequence ((stream file-cache-stream) sequence &optional (start 0) end)
-  (assert (member (direction stream) '(:output :io)))
-  (let* ((end (or end (file-length* stream)))
-         (end2 (min end (length sequence))))
-    (when (> end2 (file-length* stream))
-      (setf (buffer stream) (adjust-array (buffer stream) end2 :initial-element 0)))
-    (replace (buffer stream) sequence
-             :start1 (file-position* stream)
-             :end1 (+ (file-position* stream) (- end start))
-             :start2 start
-             :end2 end2)
-    (incf (file-position* stream) (- end2 start))))
+  (assert (input-stream-p stream))
+  (if (>= (file-%position stream) (file-%length stream))
+      :eof
+      (multiple-value-bind (block-n block-offset)
+          (floor (file-%position stream) (file-%block-size stream))
+        (when (/= block-n (file-%block-n stream))
+          (finish-output stream)
+          (setf (file-%block-n stream) block-n
+                (file-%buffer stream) (read-file-block stream block-n)))
+        (prog1 (aref (file-%buffer stream) block-offset)
+          (incf (file-%position stream))))))
+
+;; TODO
+;; (defmethod mezzano.gray:stream-write-sequence ((stream file-cache-stream) sequence &optional (start 0) end)
+;;   (assert (output-stream-p stream))
+;;   (let* ((end (or end (file-%length stream)))
+;;          (end2 (min end (length sequence))))
+;;     (when (> end2 (file-%length stream))
+;;       (setf (file-%buffer stream) (adjust-array (file-%buffer stream) end2 :initial-element 0)))
+;;     (replace (file-%buffer stream) sequence
+;;              :start1 (file-%position stream)
+;;              :end1 (+ (file-%position stream) (- end start))
+;;              :start2 start
+;;              :end2 end2)
+;;     (incf (file-%position stream) (- end2 start))))
 
 (defmethod mezzano.gray:stream-read-sequence ((stream file-cache-stream) sequence &optional (start 0) end)
-  (assert (member (direction stream) '(:input :io)))
-  (let* ((end (or end (length sequence)))
-         (end2 (min end (file-length* stream))))
-    (replace sequence (buffer stream) :start1 start :end1 end :start2 0 :end2 end2)
-    end2))
+  (assert (input-stream-p stream))
+  (let* ((buffer-length (file-%block-size stream))
+         (end1 (or end (length sequence)))
+         (end2 (+ (file-%position stream) (min end1 (file-%length stream)))))
```

Comment:
I'm not sure what this is doing. If it's supposed to be the end file position, shouldn't it be `(min (+ (file-%position stream) (- end1 start)) (file-%length stream))`?